ambry-router/src/main/java/com/github/ambry/router/ParanoidDurabilityOperationTracker.java

```diff
@@ -0,0 +1,405 @@
+ * require strong durability guarantees.
+ *
+ */
+public class ParanoidDurabilityOperationTracker extends SimpleOperationTracker {
+  private static final Logger logger = LoggerFactory.getLogger(ParanoidDurabilityOperationTracker.class);
+  protected final int localReplicaSuccessTarget;
+  protected int remoteReplicaSuccessTarget = 0;
+  protected final int localReplicaParallelism;
+  protected final int remoteReplicaParallelism;
+  protected int localInflightCount = 0;
+  protected int remoteInflightCount = 0;
+  protected int localReplicaSuccessCount = 0;
+  protected int remoteReplicaSuccessCount = 0;
+  private Map<String, LinkedList<ReplicaId>> replicaPoolByDc = new HashMap<>();
+  private Iterator<ReplicaId> replicaByDcIterator;
+  private Map<String, Integer> inflightCountByDc;
+  private Map<String, Integer> failedCountByDc;
+  private Map<String, Integer> replicaSuccessCountByDc;
+  private Map<String, Integer> disabledCountByDc;
+  private Map<String, Integer> replicaInPoolOrFlightCountByDc;
+  private Map<String, Integer> totalReplicaCountByDc;
+  private Set<String> allDatacenters;
+  private List<String> remoteDatacenters;
+
+  ParanoidDurabilityOperationTracker(RouterConfig routerConfig, PartitionId partitionId, String originatingDcName,
+      NonBlockingRouterMetrics routerMetrics) {
+    super(routerConfig, RouterOperation.PutOperation, partitionId, originatingDcName, true, routerMetrics);
+
+    localReplicaParallelism = routerConfig.routerPutLocalRequestParallelism;
+    remoteReplicaParallelism = routerConfig.routerPutRemoteRequestParallelism;
+
+    // In SimpleOperationTracker, what do we actually do with eligibleReplicas, and do we really need this variable here?
+    // Note this statement gets the list of eligible replicas in the LOCAL DC only.
+    // Tommy: We should not use the local replicas here, we should just get a list of ALL the replicas like we do with
+    //        the allReplicas variable just below here.
+    List<ReplicaId> localEligibleReplicas = getEligibleReplicas(datacenterName, EnumSet.of(ReplicaState.STANDBY, ReplicaState.LEADER));
+    List<ReplicaId> offlineReplicas = new ArrayList<>();
+    // This is a list of all replicas for this partition. This comes from replica information updated by us in helix
+    // Data node configs via bootstrap tool and not external view. This would be same as ideal view.
+    List<ReplicaId> allReplicas = new ArrayList<>(partitionId.getReplicaIds());
+
+    localReplicaSuccessTarget =
+        routerConfig.routerGetEligibleReplicasByStateEnabled ? Math.max(localEligibleReplicas.size() - 1, routerConfig.routerPutSuccessTarget) : routerConfig.routerPutSuccessTarget;
+    remoteReplicaSuccessTarget = routerConfig.routerPutRemoteSuccessTarget;
+
+    // Tommy: End of switch statement operations.
+
+    addReplicasToPool(allReplicas, offlineReplicas, true, false);
+
+
```

Comment:
We should initialize a ParanoidDurabilityTrackerIterator here and return it in `getReplicaIterator`.